### PR TITLE
[PF-17] SDUI OEmbed Block

### PR DESCRIPTION
**Pull request by stevestreza-ksr**
# 📲 What

Adds the OEmbed block (aka the YouTube player block) to the RichTextView.

# 🤔 Why

This is what powers the YouTube player. I'm sure someone is using it for something else, too.

# 🛠 How

Added the OEmbed block view the same way we have other views, added a bunch of AI-generated tests for it, and integrated it into the RichTextView.

The ExternalSourceViewElementCell was cribbed from to figure out some edge cases (this is what we currently use for project story today).

# 👀 See

Screen Recording 2026-05-28 at 18 20 47

### CODE REVIEW

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import KDS
+import SwiftUI
+import WebKit
+
+struct OEmbedBlock: View {
+  var oembed: RichTextElement.Oembed
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift]: This is the tiniest of nits, but: is there a canonical capitalization? Since it's `oembed` here and `OEmbed` in the View, I'd expect it to be `oEmbed` and `OEmbedBlock` or `oembed` and `OembedBlock`

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+// MARK: - OEmbedWebView
+
+private struct OEmbedWebView: UIViewRepresentable {
+  let url: URL
+  var onWebViewCreated: ((WKWebView) -> Void)?
+
+  /// Referrer sent with embed requests so providers can validate the source domain.
+  static let referrer = "https://www.kickstarter.com"
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift]: Nit: Should this always be `kickstarter.com`? Or should it respect the app's environment and flip to `staging.kickstarter.com` if necessary?

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  func makeCoordinator() -> Coordinator {
+    Coordinator()
+  }
+
+  func makeUIView(context: Context) -> WKWebView {
+    let configuration = WKWebViewConfiguration()
+    configuration.allowsInlineMediaPlayback = true
+    configuration.suppressesIncrementalRendering = true
+    configuration.mediaTypesRequiringUserActionForPlayback = .all
+    configuration.applicationNameForUserAgent = "Kickstarter-iOS"
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift:68]: Same question here - should this change when it's the beta app?

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+      _: WKWebView,
+      createWebViewWith _: WKWebViewConfiguration,
+      for navigationAction: WKNavigationAction,
+      windowFeatures _: WKWindowFeatures
+    ) -> WKWebView? {
+      let targetsNewWindow = navigationAction.targetFrame == nil
+        || navigationAction.targetFrame?.isMainFrame == false
+
+      if targetsNewWindow, let url = navigationAction.request.url {
+        UIApplication.shared.open(url)
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift:111]: What's a situation where this would happen? Like tapping on an embedded callout in a video?

File: ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+
+    let view = block
+      .environment(\.richTextStyle, richTextStyle(.light))
+      .frame(width: 300, height: 200)
+
+    ViewHosting.host(view: view)
+    defer { ViewHosting.expel() }
+    wait(for: [expectation], timeout: 1.0)
+
+    let webView = try XCTUnwrap(capturedWebView, "Expected the onWebViewCreated callback to fire.")
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift:167]: Nit: Is `onWebViewCreated` just for testing? If so can you add a comment to that effect?

File: ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift [hunk 6]
```diff
@@ -0,0 +1,439 @@
...
+      let view = oembedBlock(oembed: oembed, colorScheme: colorScheme)
+      let block = try view.inspect().find(OEmbedBlock.self).actualView()
+
+      XCTAssertEqual(
+        block.oembed.title,
+        "Styled Embed",
+        "OEmbedBlock.oembed.title should be preserved for color scheme \(colorScheme)."
+      )
+      XCTAssertNoThrow(
+        try view.inspect().find(viewWithAccessibilityLabel: "Styled Embed"),
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift]: This test seems redundant/tautological - is it actually testing anything about the color scheme?

File: ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift [hunk 7]
```diff
@@ -0,0 +1,439 @@
...
+    )
+    XCTAssertEqual(
+      frame.height,
+      containerHeight,
+      "Expected OEmbedBlock height to respect container height."
+    )
+  }
+}
+
+// MARK: - Helpers
```
Comment (amy-at-kickstarter) [inline ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift:363]: These view inspection tests are pretty neat - but they reminded me of something. Do we have a vanilla screenshot test that covers all the SDUI elements? Might be nice to have a snapshot of the sample page, even if we can't do a full exercise like the XXL font sizes right now.

**Review by amy-at-kickstarter — COMMENTED** [review]
Some questions before I approve!

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift (hunk 3 above)
Comment (stevestreza-ksr) [reply ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift:68]: We don't currently do this in [WebViewController](https://github.com/kickstarter/ios-oss/blob/ccbae3ef8c1238825588ab6afad2af5d374012bb/Kickstarter-Framework/Sources/Kickstarter-Framework/Kickstarter-iOS/Library/WebViewController.swift#L20) or [ExternalSourceViewElementCell](https://github.com/kickstarter/ios-oss/blob/ccbae3ef8c1238825588ab6afad2af5d374012bb/Kickstarter-Framework/Sources/Kickstarter-Framework/Kickstarter-iOS/Features/ProjectPage/Views/Cells/ExternalSourceViewElementCell.swift#L14), so I'm going to stick with this string for now.

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift (hunk 1 above)
Comment (stevestreza-ksr) [reply ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift]: Technically it's oEmbed, but I'll unify them as OEmbed, good note

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift (hunk 2 above)
Comment (stevestreza-ksr) [reply ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift]: We do switch this presently, but we don't have access to AppEnvironment which is all in Library. Maybe we could start pulling some way of abstracting that into an API that could be accessible from SwiftUI inside of a module like this.

**Comment by scottkicks** [comment]
More or less the same callouts as Amy 😅

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift (hunk 2 above)
Comment (amy-at-kickstarter) [reply ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift]: Instead of trying to untangle `AppEnvironment` from Library, what about just adding this as an additional `@Environment` variable? e.g.

```
struct OEmbedBlock: View {
  @Environment(\.referrer) var referrer: String
```

File: ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift (hunk 4 above)
Comment (stevestreza-ksr) [reply ServerDrivenUI/Sources/ServerDrivenUI/Blocks/OEmbedBlock.swift:111]: Yeah, like if you tap the title on a YouTube video, it opens the URL.

This is the exact same behavior as in ExternalSourceViewElementCell.

File: ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift (hunk 7 above)
Comment (stevestreza-ksr) [reply ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift:363]: Not specifically because loading images, videos, and OEmbeds into a snapshot test is tough.

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift (hunk 6 above)
Comment (stevestreza-ksr) [reply ServerDrivenUI/Tests/ServerDrivenUITests/Blocks/OEmbedBlockTests.swift]: No, removed

**Review by amy-at-kickstarter — APPROVED** [review]
